File: Python_implementation/emisorCRC.py
from view import *

class Emisor():

    def __init__(self, userInfo):
        userInfo = userInfo

        self.trama = []

        for char in userInfo:
            if char == '1':
                self.trama.append(True)

            elif char == '0':
                self.trama.append(False)

        # Save a copy of the original before padding
        self.tramaOriginal = []
        for t in self.trama:
            self.tramaOriginal.append(t)

        # padding
        for i in range(31):
            self.trama.append(False)

        print("\ncon padding")
        printTrama(self.trama)


        # generate polinomio
        self.polinomio = [
            True, False, True, True, False, True, False, True, 
            False, False, False, True, False, True, False, True,
            False, False, False, False, True, True, True, False, 
            True, False, False, True, True, False, True, False
        ]

        self.CRC()

    def CRC(self):
        actual = []

        # Initialize actual
        for i in range(len(self.polinomio)):
            actual.append(self.trama[i] ^ self.polinomio[i])

        print()
        printTrama(self.trama)
        printTrama(self.polinomio)
        printTrama(actual)

        nextB = len(self.polinomio)

        while(nextB < len(self.trama)):

            # remove first 0 and add next value
            actual = actual[1:]
            actual.append(self.trama[nextB])
            nextB += 1

            # check if the first digit is 1
            if (actual[0]):

                print()
                printTrama(actual)
                printTrama(self.polinomio)

                temp = [actual[i] ^ self.polinomio[i] for i in range(len(self.polinomio))]

                printTrama(temp)

                actual = []
                actual = [x for x in temp]

            else:
                print()
                printTrama(actual)

        print("\nResultado emisor")
        self.result = [x for x in self.tramaOriginal]
        actual = actual[1:] # remove first 0
        for x in actual:
            self.result.append(x)
        printTrama(self.result)

    def get_response(self):
        res = ""

        for b in self.result:
            if (b): res += "1"
            else: res += "0"

        return res
        
    
        


        

        

        
        

    

File: Java_implementation/EmisorCRC.java
import java.util.ArrayList;
import java.util.Random;

public class EmisorCRC {

    Random rand = new Random();
    View view = new View();

    ArrayList<Boolean> trama = new ArrayList<>();
    ArrayList<Boolean> tramaOriginal;
    ArrayList<Boolean> polinomio;
    ArrayList<Boolean> result;

    public EmisorCRC(String input) {

        for (char c: input.toCharArray()) {
            if (c == '1') {
                trama.add(true);
            }
            else if (c == '0') trama.add(false);
        }

        // save a copy of the original before padding
        tramaOriginal = new ArrayList<>(trama);

        // Padding
        // while (trama.size() % 32 != 0 ) {
            for ( int i = 0; i < 31; i ++  ) {
                trama.add(false);
            }

        System.out.println("\ncon padding");
        view.printTrama(trama);

        // generate polinomio
        polinomio = new ArrayList<>();
        Boolean[] pol = {
            true, false, true, true, false, true, false, true, 
            false, false, false, true, false, true, false, true,
            false, false, false, false, true, true, true, false, 
            true, false, false, true, true, false, true, false
        };
        for (boolean b: pol) polinomio.add(b);

        System.out.println("\npolinomio");
        view.printTrama(polinomio);

        // excecute algorithm
        CRC();

    }

    private void CRC() {

        ArrayList<Boolean> actual = new ArrayList<>();

        // initialize by copying the first bits of the trama
        for (int i = 0; i < polinomio.size(); i++) {
            actual.add(trama.get(i) ^ polinomio.get(i));
        }
        System.out.println();
        view.printTrama(trama);
        view.printTrama(polinomio);
        view.printTrama(actual);

        int nextB = polinomio.size();

        while(nextB < trama.size()) {

            // remove the first 0 and add next bit
            
            actual.remove(0);
            actual.add(trama.get(nextB));
            nextB++;
            
            // check if the first digit is 0            
            if (actual.get(0)) {

                System.out.println();
                view.printTrama(actual);
                view.printTrama(polinomio);

                // oparate normally
                ArrayList<Boolean> temp = new ArrayList<>();
                for (int i = 0; i < polinomio.size(); i ++) {
                    temp.add(actual.get(i) ^ polinomio.get(i));
                }

                view.printTrama(temp);

                // replace
                actual = new ArrayList<>(temp);
            }

            else {
                System.out.println();
                view.printTrama(actual);
            }

        }

        System.out.println("\nResultado");
        result = new ArrayList<>(tramaOriginal);
        actual.remove(0); // remove first 0
        result.addAll(actual);
        view.printTrama(result);

    }

    public ArrayList<Boolean> getPolinomio() {
        return polinomio;
    }

    public String get_response() {
        String res = "";
        for (boolean b: result) {
            if (b) res += "1";
            else res += "0";
        }
        return res;
    }

    public boolean getParidad() {
        // Calculate paridad
        int cantOnes = 0;
        for (boolean b: trama) {
            if (b) cantOnes ++;
        }

        if (cantOnes % 2 == 0) return true;
        else return false;
    }
    
    
}
